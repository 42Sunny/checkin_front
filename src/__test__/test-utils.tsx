// test-utils.jsx
import { configureStore } from "@reduxjs/toolkit";
import { Queries, render, RenderResult } from "@testing-library/react";
import { createMemoryHistory } from "history";
import React, { ReactElement } from "react";
import { Provider } from "react-redux";
import { Router } from "react-router-dom";
// Import your own reducer
import renderer from "react-test-renderer";
import appStore from "../redux/configureStore";
import rootReducer, { RootState } from "../redux/modules";
import { initialState as clusterInitState } from "../redux/modules/cluster";
import { initialState as userInitState } from "../redux/modules/user";

const makeCustomStore = ({
  cluster,
  user,
}: {
  cluster?: typeof clusterInitState;
  user?: typeof userInitState;
} = {}) => {
  return {
    clusterReducer: { ...clusterInitState, ...cluster },
    userReducer: { ...userInitState, ...user },
  };
};

const customRender = (
  ui: ReactElement,
  preloadedState: RootState = makeCustomStore(),
  { store = configureStore({ reducer: rootReducer, preloadedState }), ...renderOptions } = {},
  history = createMemoryHistory(),
): RenderResult<Queries, HTMLElement> => {
  const Wrapper: any = ({ children }: { children: ReactElement }) => {
    return (
      <Provider store={store}>
        <Router history={history}>{children}</Router>
      </Provider>
    );
  };
  return render(ui, { wrapper: Wrapper, ...renderOptions });
};

const createWithWrapper = (children: ReactElement) =>
  renderer.create(
    <Provider store={appStore}>
      <Router history={createMemoryHistory()}>{children}</Router>
    </Provider>,
  );

// re-export everything
export * from "@testing-library/react";
// override render method
export { customRender as render, createWithWrapper, makeCustomStore };
