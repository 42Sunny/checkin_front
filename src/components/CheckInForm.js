import { useCallback } from 'react';
import { useSelector, useDispatch, shallowEqual } from 'react-redux';
import Checkbox from '../components/Checkbox';
import TextField from '@material-ui/core/TextField';

import { checkLists } from '../utils/notice';

import { setCardNum } from '../redux/modules/user';
import '../styles/CheckInForm.css';

const CheckInForm = ({ checkAll, setCheckAll, checkStatus, setCheckStatus, isFold, setIsFold }) => {
  const dispatch = useDispatch();
  const { cardNum } = useSelector(
    state => ({
      cardNum: state.user.cardNum
    }),
    shallowEqual
  );

  const handleCheckAll = useCallback(
    e => {
      const isChecked = e.target.checked;
      setCheckAll(isChecked);
      setCheckStatus([isChecked, isChecked, isChecked]);
      setIsFold(true);
    },
    [setCheckAll, setCheckStatus, setIsFold]
  );

  // const handleClick = () => {
  //   setIsFold(state => !state);
  // };
  return (
    <>
      <div id='check-in-form-wrapper'>
        <input
          id='allCheck'
          style={{ fontSize: '1em', margin: '5px' }}
          type='checkbox'
          checked={checkAll}
          onChange={handleCheckAll}
        />
        <label for='allCheck' style={{ fontSize: '1em' }}>
          <span>모두 동의</span>
          <span style={{ color: 'red', fontWeight: 'bold' }}>*</span>
        </label>
        {/* {!isFold && ( */}
        <div>
          {checkLists.map((checkList, idx) => (
            <Checkbox
              key={idx}
              idx={idx}
              text={checkList}
              checkStatus={checkStatus}
              setCheckStatus={setCheckStatus}
            />
          ))}
        </div>
        {/* )} */}
      </div>
      <div style={{ marginBottom: '10px' }}>
        <TextField
          id='standard-basic'
          label='Card Number'
          value={cardNum}
          onChange={e => {
            dispatch(setCardNum(e.target.value));
          }}
        />
      </div>
    </>
  );
};
export default CheckInForm;
