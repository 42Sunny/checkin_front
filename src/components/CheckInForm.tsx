import React, { useEffect, useState } from "react";
import classes from "../styles/CheckInForm.module.css";
import Button from "./Button";
import List from "./List";

const checkLists = [
  "발열 체크시 37.5도 이하인 것을 확인했습니다.",
  "임시 출입카드 분실 시 분실 비용이 발생하는 것을 확인했습니다.",
  "마스크를 반드시 상시 착용하고 방역수칙을 준수할 것을 약속하며, 모든 설문을 이상없이 작성했음을 확인합니다.",
];

interface CardInputProps {
  cardNum: string;
  handleCardNumberChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

const CardInput: React.FC<CardInputProps> = ({ cardNum, handleCardNumberChange }) => (
  <div className={classes["cardNumber-wrapper"]}>
    <input
      className={classes.cardNumber}
      type='number'
      min={1}
      value={cardNum}
      inputMode='numeric' /* 숫자형 키패드 */
      placeholder='카드 번호'
      onChange={handleCardNumberChange}
    />
  </div>
);

interface CheckListProps {
  checkAll: boolean;
  handleCheckAll: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

const CheckList: React.FC<CheckListProps> = ({ checkAll, handleCheckAll }) => (
  <div className={classes["check-in-form-wrapper"]}>
    <label htmlFor='allCheck' className={classes.allCheck}>
      <input
        id='allCheck'
        type='checkbox'
        checked={checkAll}
        onChange={handleCheckAll}
        className={classes.allCheckBox}
      />
      <span>클러스터 이용 약관에 모두 동의합니다.</span>
      <span className={classes.asterisk}>*</span>
    </label>
    <ul className={classes["check-list-wrapper"]}>
      {checkLists.map((checkList) => (
        <List key={checkList} text={checkList} />
      ))}
    </ul>
  </div>
);

interface IProps {
  handleCheckIn: (cardNum: string) => (e: React.FormEvent<HTMLFormElement>) => Promise<boolean>;
}

// const waitingNotice = [
//   " 입장 안내 알림을 받은 후로 10분 이내에 체크인을 완료하지 않을 시에 대기가 자동 취소됨을 확인합니다.",
// ];

const CheckInForm: React.FC<IProps> = ({ handleCheckIn }) => {
  const [cardNum, setCardNum] = useState("");
  const [checkAll, setCheckAll] = useState(false);
  const [readySubmit, setReadySubmit] = useState(false);

  const handleCheckAll = (e: React.ChangeEvent<HTMLInputElement>) => {
    setCheckAll(e.target.checked);
  };

  const handleCardNumberChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setCardNum(e.target.value);
  };

  const checkSubmitCondition = () => {
    if (cardNum && checkAll) setReadySubmit(true);
    else setReadySubmit(false);
  };

  useEffect(() => {
    checkSubmitCondition();
    return () => {
      setReadySubmit(false);
    };
  });

  return (
    <form className={classes.form} onSubmit={handleCheckIn(cardNum)}>
      <CheckList checkAll={checkAll} handleCheckAll={handleCheckAll} />
      <CardInput cardNum={cardNum} handleCardNumberChange={handleCardNumberChange} />
      <Button type='submit' className={classes.submitBtn} text='CHECK IN' disabled={!readySubmit} />
    </form>
  );
};

export default CheckInForm;
