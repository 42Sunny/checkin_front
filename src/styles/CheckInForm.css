#check-in-form-wrapper {
  background-color: rgb(244, 244, 245);
  width: 90%;
  border-radius: 3px;
  padding: 1rem 0;
}

#cardNumber {
  border-style: none;
  border-bottom: 1px solid rgba(105, 105, 105, 0.2);
  padding: 0.3rem;
  font-size: 1.2rem;
  text-align: center;
}

#cardNumber::placeholder {
  color: rgba(105, 105, 105, 0.2);
}

/* Chrome, Safari, Edge, Opera */
input::-webkit-outer-spin-button,
input::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

/* Firefox */
input[type='number'] {
  -moz-appearance: textfield;
}
