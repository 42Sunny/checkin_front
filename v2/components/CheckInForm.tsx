import React, { useCallback, useEffect, useState } from "react";
import classes from "../styles/components/CheckInUi.module.css";
import Button from "./common/Button";
import Checkbox from "./common/Checkbox";

interface CardInputProps {
  cardNum: string;
  handleCardNumberChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

const CardInput: React.FC<CardInputProps> = ({ cardNum, handleCardNumberChange }) => (
  <input
    tabIndex={0}
    className={classes.cardNumber}
    type='number'
    min={1}
    step={1}
    value={cardNum}
    inputMode='decimal' /* 숫자형 키패드 */
    placeholder='카드 번호'
    onChange={handleCardNumberChange}
  />
);

interface CheckInputProps {
  handleModalOpen: () => void;
  isChecked: boolean;
  handleCheck: (e: React.ChangeEvent<HTMLInputElement>) => void;
}
const CheckInput: React.FC<CheckInputProps> = ({ handleCheck, handleModalOpen, isChecked }) => (
  <div className={classes.checkWrap}>
    <Checkbox text='방역수칙에 동의하고 입장합니다.' checked={isChecked} onChange={handleCheck} />
    <button type='button' className={classes.checkDetail} onClick={handleModalOpen}>
      자세히
    </button>
  </div>
);

interface CheckInFormProps {
  handleCheckIn: (cardNum: string) => (e: React.FormEvent<HTMLFormElement>) => Promise<boolean>;
  handleModalOpen: () => void;
}
const CheckInForm: React.FC<CheckInFormProps> = ({ handleCheckIn, handleModalOpen }) => {
  const [cardNum, setCardNum] = useState("");
  const [isChecked, setIsChecked] = useState(false);
  const [readySubmit, setReadySubmit] = useState(false);

  const handleCheck = (e: React.ChangeEvent<HTMLInputElement>) => {
    setIsChecked(e.target.checked);
  };

  const handleCardNumberChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setCardNum(e.target.value);
  };

  const checkSubmitCondition = useCallback(() => {
    if (cardNum && isChecked) setReadySubmit(true);
    else setReadySubmit(false);
  }, [cardNum, isChecked]);

  useEffect(() => {
    checkSubmitCondition();
  }, [checkSubmitCondition]);

  return (
    <form className={classes["check-in-form"]} onSubmit={handleCheckIn(cardNum)}>
      <CardInput cardNum={cardNum} handleCardNumberChange={handleCardNumberChange} />
      <CheckInput
        handleModalOpen={handleModalOpen}
        isChecked={isChecked}
        handleCheck={handleCheck}
      />
      <Button
        className={classes["check-in-button"]}
        type='submit'
        text='CHECK IN'
        disabled={!readySubmit}
      />
    </form>
  );
};

export default CheckInForm;
