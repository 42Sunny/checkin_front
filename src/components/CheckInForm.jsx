/* eslint-disable react/no-array-index-key */
/* eslint-disable jsx-a11y/label-has-associated-control */
// TODO: fix lint rule
import React, { useCallback } from "react";

import { useSelector, useDispatch, shallowEqual } from "react-redux";
import Checkbox from "./Checkbox";
import { checkLists } from "../utils/notice";
import { setCardNum } from "../redux/modules/user";
import "../styles/CheckInForm.css";

// interface IProps {
//   checkAll: boolean;
//   setCheckAll: string;
//   checkStatus: string;
//   setCheckStatus: string;
//   isFold: string;
//   setIsFold: string;
// }

const CheckInForm = ({ checkAll, setCheckAll, checkStatus, setCheckStatus, isFold, setIsFold }) => {
  const dispatch = useDispatch();
  const { cardNum } = useSelector(
    (state) => ({
      cardNum: state.user.cardNum,
    }),
    shallowEqual,
  );

  const handleCheckAll = useCallback(
    (e) => {
      const isChecked = e.target.checked;
      setCheckAll(isChecked);
      setCheckStatus([isChecked, isChecked, isChecked]);
      setIsFold(true);
    },
    [setCheckAll, setCheckStatus, setIsFold],
  );

  // const handleClick = () => {
  //   setIsFold(state => !state);
  // };
  return (
    <>
      <div id='check-in-form-wrapper'>
        <input
          id='allCheck'
          style={{ fontSize: "1em", margin: "5px" }}
          type='checkbox'
          checked={checkAll}
          onChange={handleCheckAll}
        />
        <label htmlFor='allCheck' style={{ fontSize: "1em" }}>
          <span>모두 동의</span>
          <span style={{ color: "red", fontWeight: "bold" }}>*</span>
        </label>
        {/* {!isFold && ( */}
        <div>
          {checkLists.map((checkList, idx) => (
            <Checkbox
              key={idx}
              idx={idx}
              text={checkList}
              checkStatus={checkStatus}
              setCheckStatus={setCheckStatus}
            />
          ))}
        </div>
        {/* )} */}
      </div>
      <div style={{ margin: "10px 0" }}>
        <input
          id='cardNumber'
          type='number'
          value={cardNum}
          inputMode='numeric' /* 숫자형 키패드 */
          placeholder='카드 번호'
          onChange={(e) => {
            dispatch(setCardNum(e.target.value));
          }}
        />
      </div>
    </>
  );
};
export default CheckInForm;
