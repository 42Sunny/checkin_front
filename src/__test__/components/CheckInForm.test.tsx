import { fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { useEffect } from "react";
import CheckInForm from "../../components/CheckInForm";
import { act, screen, createWithWrapper, render } from "../test-utils";

describe("src/components/CheckInForm.tsx", () => {
  const checkInFormProps: {
    handleCheckIn: any;
    handleModalOpen: any;
  } = {
    handleCheckIn: () => {},
    handleModalOpen: () => {},
  };

  beforeEach(() => {
    const curriedCheckIn = jest.fn((e) => e.preventDefault());
    checkInFormProps.handleCheckIn = jest.fn(() => curriedCheckIn);
    checkInFormProps.handleModalOpen = jest.fn();
  });

  it("matches snapshot", () => {
    const component = createWithWrapper(<CheckInForm {...checkInFormProps} />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("renders with button disabled", () => {
    render(<CheckInForm {...checkInFormProps} />);

    const buttonEL = screen.getByText(/check in/i);
    expect(buttonEL).toBeDisabled();
  });

  it("card input's initial value is  '' ", () => {
    render(<CheckInForm {...checkInFormProps} />);
    const cardInputEl = screen.getByPlaceholderText(/카드 번호/i) as HTMLInputElement;

    expect(cardInputEl.value).toBe("");
  });

  it("card input doesn't accept non-numeric value", () => {
    render(<CheckInForm {...checkInFormProps} />);
    const cardInputEl = screen.getByPlaceholderText(/카드 번호/i) as HTMLInputElement;
    userEvent.type(cardInputEl, "a");

    expect(cardInputEl.value).toBe("");
  });

  it("card input accept numeric value", () => {
    render(<CheckInForm {...checkInFormProps} />);

    const cardInputEl = screen.getByPlaceholderText(/카드 번호/i) as HTMLInputElement;
    userEvent.type(cardInputEl, "1241");

    expect(cardInputEl.value).toBe("1241");
  });

  it("checkbox's initial value is not checked", () => {
    render(<CheckInForm {...checkInFormProps} />);

    const checkboxInputEl = screen.getByRole("checkbox") as HTMLInputElement;

    expect(checkboxInputEl.checked).toBe(false);
  });

  it("checkbox is clicked", () => {
    render(<CheckInForm {...checkInFormProps} />);

    const checkboxInputEl = screen.getByRole("checkbox") as HTMLInputElement;
    userEvent.click(checkboxInputEl);

    expect(checkboxInputEl.checked).toBe(true);
  });

  it("when 자세히 button is clicked handleModalOpen gets called", () => {
    render(<CheckInForm {...checkInFormProps} />);

    const 자세히btnEl = screen.getByText(/자세히/i) as HTMLButtonElement;

    userEvent.click(자세히btnEl);

    expect(checkInFormProps.handleModalOpen).toHaveBeenCalledTimes(1);
  });

  it("user types card Input and checks checkbox button gets enabled", () => {
    render(<CheckInForm {...checkInFormProps} />);
    const checkboxInputEl = screen.getByRole("checkbox") as HTMLInputElement;
    const cardInputEl = screen.getByPlaceholderText(/카드 번호/i) as HTMLInputElement;
    const buttonEl = screen.getByText(/check in/i);

    userEvent.click(checkboxInputEl);
    userEvent.type(cardInputEl, "123");

    expect(buttonEl).toBeEnabled();
  });

  it("when submit condition is fulfilled and user clicks submitBtn handleCheckIn gets called", () => {
    render(<CheckInForm {...checkInFormProps} />);
    const checkboxInputEl = screen.getByRole("checkbox") as HTMLInputElement;
    const cardInputEl = screen.getByPlaceholderText(/카드 번호/i) as HTMLInputElement;
    const buttonEl = screen.getByText(/check in/i);

    userEvent.click(checkboxInputEl);
    userEvent.type(cardInputEl, "123");
    userEvent.click(buttonEl);
    expect(checkInFormProps.handleCheckIn()).toBeCalledTimes(1);
  });

  it("when submit condition is not fulfilled and user clicks submitBtn handleCheckIn doesn't gets called", () => {
    render(<CheckInForm {...checkInFormProps} />);
    const buttonEl = screen.getByText(/check in/i);

    userEvent.click(buttonEl);

    expect(checkInFormProps.handleCheckIn()).toBeCalledTimes(0);
  });
});
