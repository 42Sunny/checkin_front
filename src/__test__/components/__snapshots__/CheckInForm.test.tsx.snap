// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/CheckInForm.tsx matches snapshot 1`] = `
<form
  className="check-in-form"
  onSubmit={[MockFunction]}
>
  <input
    className="cardNumber"
    inputMode="decimal"
    min={1}
    onChange={[Function]}
    placeholder="카드 번호"
    step={1}
    tabIndex={0}
    type="number"
    value=""
  />
  <div
    className="checkWrap"
  >
    <label
      className="text"
    >
      <input
        checked={false}
        className="checkbox"
        onChange={[Function]}
        type="checkbox"
      />
      방역수칙에 동의하고 입장합니다.
    </label>
    <button
      className="checkDetail"
      onClick={[MockFunction]}
      type="button"
    >
      자세히
    </button>
  </div>
  <button
    className="button green"
    color="green"
    disabled={true}
    type="submit"
  >
    CHECK IN
  </button>
</form>
`;
